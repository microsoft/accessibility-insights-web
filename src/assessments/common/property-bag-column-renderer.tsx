// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InstanceTableRow } from 'assessments/types/instance-table-data';
import { ColumnValueBag } from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { isEmpty } from 'lodash';
import * as React from 'react';

import { DictionaryStringTo } from 'types/common-types';

export function propertyBagColumnRenderer<TPropertyBag extends ColumnValueBag>(
    item: InstanceTableRow<TPropertyBag>,
    configs: PropertyBagColumnRendererConfig<TPropertyBag>[],
): JSX.Element {
    const mapper = (config: PropertyBagColumnRendererConfig<TPropertyBag>, index: number) => {
        if (item.instance.propertyBag == null) {
            return null;
        }
        const value = item.instance.propertyBag[config.propertyName];
        if (value == null && config.defaultValue == null) {
            return null;
        }

        return render(config, value, index);
    };

    const render = (
        config: PropertyBagColumnRendererConfig<TPropertyBag>,
        value: any,
        index: number,
    ) => {
        return (
            <div key={`property-${index}`} className="property-bag-div">
                <span className="display-name">{`${config.displayName}: `}</span>
                {renderValue(config, value)}
            </div>
        );
    };

    const renderValue = (config: PropertyBagColumnRendererConfig<TPropertyBag>, value: any) => {
        if (config.expand) {
            return renderProperties(config, value);
        }

        return <React.Fragment>{value || config.defaultValue}</React.Fragment>;
    };

    const renderProperties = (
        config: PropertyBagColumnRendererConfig<TPropertyBag>,
        propertyMap: DictionaryStringTo<string>,
    ) => {
        if (isEmpty(propertyMap)) {
            return <React.Fragment>{config.defaultValue.toString()}</React.Fragment>;
        }

        return Object.keys(propertyMap).map(key => {
            return (
                <div key={key} className="expanded-property-div">
                    {renderInnerKeyValue(key, propertyMap[key])}
                </div>
            );
        });
    };

    const renderInnerKeyValue = (key: string, value: string) => {
        if (value === null) {
            return <span className="display-name">{`${key}`}</span>;
        } else {
            return (
                <React.Fragment>
                    <span className="display-name">{`${key}: `}</span>
                    {value}
                </React.Fragment>
            );
        }
    };

    return <div className="property-bag-container">{configs.map(mapper)}</div>;
}
