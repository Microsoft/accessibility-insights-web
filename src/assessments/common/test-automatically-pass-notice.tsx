// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from 'react';
import * as Markup from '../markup';

const TestAutomaticallyPassedNotice = (
    <p>
        <Markup.Emphasis>
            Note: If no matching/failing instances are found, this requirement will automatically be marked as pass.
        </Markup.Emphasis>
    </p>
);

export default TestAutomaticallyPassedNotice;
