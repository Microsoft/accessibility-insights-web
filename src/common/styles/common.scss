// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
