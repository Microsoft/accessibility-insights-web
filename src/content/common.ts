// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { link } from './link';
import { ContentCreator } from '../views/content/content-page';
import { GuidanceTitle as gt } from '../views/content/guidance-title';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
