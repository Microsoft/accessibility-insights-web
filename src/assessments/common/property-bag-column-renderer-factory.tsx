// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ColumnValueBag } from '../../common/types/property-bag/column-value-bag';
import { IAssessmentInstanceRowData } from '../../DetailsView/components/assessment-instance-table';
import { propertyBagColumnRenderer, PropertyBagColumnRendererConfig } from './property-bag-column-renderer';

export class PropertyBagColumnRendererFactory {
    public static get<TPropertyBag extends ColumnValueBag>(
        configs: PropertyBagColumnRendererConfig<TPropertyBag>[],
    ): (item: IAssessmentInstanceRowData<TPropertyBag>) => JSX.Element {
        return item => {
            return propertyBagColumnRenderer(item, configs);
        };
    }
}
