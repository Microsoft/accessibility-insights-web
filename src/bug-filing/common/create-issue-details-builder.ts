// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { EnvironmentInfo } from '../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../common/types/create-issue-details-text-data';
import { title } from '../../content/strings/application';
import { IssueDetailsBuilder } from './issue-details-builder';
import { MarkupFactory } from './markup-factory';

export const createIssueDetailsBuilder = (markup: MarkupFactory): IssueDetailsBuilder => {
    const getter = (environmentInfo: EnvironmentInfo, data: CreateIssueDetailsTextData): string => {
        const result = data.ruleResult;

        const text = [
            `${markup.bold('Issue')}: ${markup.snippet(result.help)} (${markup.link(result.helpUrl, result.ruleId)})`,
            markup.sectionSeparator(),
            `${markup.bold('Target application')}: ${markup.link(data.pageUrl, data.pageTitle)}`,
            markup.sectionSeparator(),
            `${markup.bold('Element path')}: ${data.ruleResult.selector}`,
            markup.sectionSeparator(),
            `${markup.bold('Snippet')}:`,
            markup.snippet(result.snippet),
            markup.sectionSeparator(),
            `${markup.bold('How to fix')}:`,
            markup.newLine(),
            markup.howToFixSection(result.failureSummary),
            markup.sectionSeparator(),
            `${markup.bold('Environment')}:`,
            environmentInfo.browserSpec,
            markup.sectionSeparator(),
            `This accessibility issue was found using ${title} ` +
                `${environmentInfo.extensionVersion} (axe-core ${environmentInfo.axeCoreVersion}), ` +
                'a tool that helps find and fix accessibility issues. Get more information & download ' +
                `this tool at ${markup.link('http://aka.ms/AccessibilityInsights')}.`,
        ].join('\n');

        return text;
    };

    return getter;
};
