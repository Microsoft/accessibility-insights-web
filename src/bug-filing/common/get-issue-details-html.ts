// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { EnvironmentInfo } from '../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../common/types/create-issue-details-text-data';
import { title } from '../../content/strings/application';
import { IssueUrlCreationUtils } from './issue-filing-url-string-utils';

const buildBodySection = (headingHtml: string, contentHtml: string): string => {
    return `<div><b>${headingHtml}</b></div><div>${contentHtml}</div><br>`;
};

const buildIssueDetails = (help: string, helpUrl: string, ruleId: string): string => {
    const helpEscapedForHtml = escape(help);
    const ruleIdEscapedForHtml = escape(ruleId);

    return `${helpEscapedForHtml}<br><a href="${helpUrl}">${ruleIdEscapedForHtml}</a>`;
};

const buildApplicationHtml = (pageTitle: string, pageUrl: string): string => {
    const pageTitleEscapedForHtml = escape(pageTitle);
    const pageUrlEscapedForUrl = encodeURI(pageUrl);
    const pageUrlEscapedForHtml = escape(pageUrl);

    return `<a href="${pageUrlEscapedForUrl}">${pageTitleEscapedForHtml}<br>${pageUrlEscapedForHtml}</a>`;
};

const buildSnippetHtml = (snippet: string): string => {
    const maxSnippetLength = 256;
    let constrainedSnippet = snippet;

    if (snippet.length > maxSnippetLength) {
        constrainedSnippet = snippet.substr(0, maxSnippetLength) + '...';
    }

    return escape(constrainedSnippet);
};

const buildHowToFixHtml = (failureSummary: string): string => {
    return escape(failureSummary)
        .replace(/\n  /g, '<br>- ')
        .replace(/\n /g, '<br> ')
        .replace(/\n/g, '<br>');
};

export const getIssueDetailsHtml = (
    stringUtils: IssueUrlCreationUtils,
    environmentInfo: EnvironmentInfo,
    data: CreateIssueDetailsTextData,
): string => {
    const body =
        buildBodySection('Issue Details', buildIssueDetails(data.ruleResult.help, data.ruleResult.helpUrl, data.ruleResult.ruleId)) +
        buildBodySection('Application', buildApplicationHtml(data.pageTitle, data.pageUrl)) +
        buildBodySection('Element Path', escape(data.ruleResult.selector)) +
        buildBodySection('Snippet', buildSnippetHtml(data.ruleResult.html)) +
        buildBodySection('How to fix', buildHowToFixHtml(data.ruleResult.failureSummary)) +
        buildBodySection(
            'To reproduce',
            `Use <a href="http://aka.ms/AccessibilityInsights">${title}</a> to investigate the issue details`,
        ) +
        buildBodySection('Environment', escape(environmentInfo.browserSpec)) +
        stringUtils.getFooter(environmentInfo);

    return body;
};
