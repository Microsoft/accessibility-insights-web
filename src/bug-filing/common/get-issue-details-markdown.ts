// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { EnvironmentInfo } from '../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../common/types/create-issue-details-text-data';
import { IssueUrlCreationUtils } from './issue-filing-url-string-utils';

export const getIssueDetailsMarkdown = (
    stringUtils: IssueUrlCreationUtils,
    environmentInfo: EnvironmentInfo,
    data: CreateIssueDetailsTextData,
): string => {
    const result = data.ruleResult;

    const text = [
        `**Issue**: \`${result.help}\` ([\`${result.ruleId}\`](${result.helpUrl}))`,
        ``,
        `**Target application**: [${data.pageTitle}](${data.pageUrl})`,
        ``,
        `**Element path**: ${data.ruleResult.selector}`,
        ``,
        `**Snippet**:`,
        ``,
        `    ${stringUtils.collapseConsecutiveSpaces(result.snippet)}`,
        ``,
        `**How to fix**:`,
        ``,
        `${stringUtils.formatAsMarkdownCodeBlock(result.failureSummary)}`,
        ``,
        `**Environment**:`,
        `${environmentInfo.browserSpec}`,
        ``,
        `====`,
        ``,
        stringUtils.getFooterContent(environmentInfo),
    ].join('\n');

    return text;
};
