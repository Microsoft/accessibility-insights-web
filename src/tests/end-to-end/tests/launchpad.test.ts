// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Page } from 'puppeteer';
import { ExtensionPuppeteerConnection } from '../common/extension-puppeteer-connection';
import { getTestResourceUrl } from '../common/test-resources';

function generateFormattedHtml(innerHTMLString: string) {
    const template = document.createElement('template');
    template.innerHTML = innerHTMLString.trim();
    return template.content.cloneNode(true);
}

describe('Adhoc Panel test', () => {
    let extensionConnection: ExtensionPuppeteerConnection;
    let targetPage: Page;
    let targetPageTabId: number;
    let popupPage: Page;

    beforeAll(async () => {
        extensionConnection = await ExtensionPuppeteerConnection.connect();

        await setupNewTargetPage();
        popupPage = await extensionConnection.newExtensionPopupPage(targetPageTabId);
        await popupPage.bringToFront();
        await dismissTelemetryPopup();
    });

    afterAll(async () => {
        await extensionConnection.tearDown();
    });

    async function setupNewTargetPage() {
        targetPage = await extensionConnection.newPage(getTestResourceUrl('all.html'));

        await targetPage.bringToFront();
        targetPageTabId = await extensionConnection.getActivePageTabId();
    }

    async function dismissTelemetryPopup() {
        await popupPage.waitForSelector('.telemetry-permission-dialog-modal');
        await popupPage.click('button.start-using-product-button');
        await popupPage.waitFor(() => !document.querySelector('.telemetry-permission-dialog-modal'));
    }

    it('test if launchpad title is there before pre', async () => {
        await popupPage.waitForSelector('.launch-pad-title');
        const launchpadTitle = await popupPage.$eval('.launch-pad-title', element => element.textContent);

        expect(launchpadTitle).toBe('Launch pad');
    });

    it('test snapshot for launchpad', async () => {
        await popupPage.waitForSelector('#new-launch-pad');
        const launchPad = await popupPage.$eval('#new-launch-pad', el => el.innerHTML);

        expect(generateFormattedHtml(launchPad)).toMatchSnapshot();
    });

    it('test if text for all the links in launchpad show properly', async () => {
        await popupPage.waitForSelector('.launch-pad-item-description');
        const launchPadItemListText = await popupPage.evaluate(() => {
            const elements = Array.from(document.querySelectorAll('.launch-pad-item-title'));
            const links = elements.map(element => {
                return element.textContent;
            });
            return links;
        });

        expect(launchPadItemListText.length).toBe(3);
        expect(launchPadItemListText).toEqual(['FastPass', 'Assessment', 'Ad hoc tools']);
    });
});
