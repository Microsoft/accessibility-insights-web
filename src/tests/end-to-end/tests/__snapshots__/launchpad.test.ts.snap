// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adhoc Panel test test snapshot for launchpad 1`] = `
<DocumentFragment>
  <header
    class="ms-Grid launch-panel-header"
  >
    <div
      class="ms-Grid-row header-title"
    >
      <div
        aria-level="1"
        class="ms-Grid-col ms-u-sm10 ms-fontColor-neutralPrimary ms-font-xl old"
        role="heading"
      >
        Accessibility Insights for Web
      </div>
      <div
        class="ms-Grid-col ms-u-sm2 feedback-collapseMenuButton-col"
      >
        <div
          class="gear-options-button-component"
        >
          <div
            class="details-view-dropdown"
          >
            <button
              class="ms-Link gear-button root-42"
            >
              <i
                aria-label="Manage Settings"
                class="gear-options-icon root-44"
                data-icon-name="Gear"
              >
                
              </i>
            </button>
          </div>
        </div>
        <button
          aria-label="Help and Feedback menu"
          class="ms-Button ms-Button--icon root-45"
          data-is-focusable="true"
          id="feedback-collapse-menu-button"
          type="button"
        >
          <div
            class="ms-Button-flexContainer flexContainer-46"
          >
            <i
              aria-hidden="true"
              class="ms-Button-icon icon-52"
              data-icon-name="GlobalNavButton"
              role="presentation"
            >
              
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      class="ms-fontColor-neutralSecondary ms-fontWeight-semilight ms-fontSize-xs"
    >
      formerly Keros for Web  |  Watch 
      <a
        aria-label="demo video"
        class="ms-Link insights-link root-53"
        href="https://msit.microsoftstream.com/video/7817f7b2-7c17-4ad8-a644-42d82ae697a2?channelId=66d47e66-d99c-488b-b9ea-98a153d2a4d4&channelName=Accessibility%20Tools"
        target="_blank"
        title="watch the 3 minute video introduction"
      >
        3-minute video
      </a>
       introduction
    </div>
  </header>
  <div
    class="ms-Grid main-section"
  >
    <main>
      <div
        aria-level="1"
        class="launch-pad-title ms-fontWeight-semibold"
        role="heading"
      >
        Launch pad
      </div>
      <hr
        class="ms-fontColor-neutralTertiaryAlt launch-pad-hr"
      />
      <div
        class="launch-pad-main-section"
      >
        <div>
          <div
            class="ms-Grid"
          >
            <div
              class="ms-Grid-row"
            >
              <div
                class="ms-Grid-col ms-sm3 popup-start-dialog-icon-circle"
              >
                <i
                  aria-hidden="true"
                  class="popup-start-dialog-icon root-44"
                  data-icon-name="Rocket"
                  role="presentation"
                >
                  
                </i>
              </div>
              <div
                class="ms-Grid-col ms-sm9"
              >
                <div
                  class="launch-pad-item-title"
                >
                  <button
                    aria-describedby="launch-pad-item-description-1"
                    class="ms-Link root-42"
                    role="link"
                  >
                    FastPass
                  </button>
                </div>
                <div
                  class="launch-pad-item-description"
                  id="launch-pad-item-description-1"
                >
                  Run two tests before check-in to find 25% of accessibility issues in less than 5 minutes.
                </div>
              </div>
            </div>
          </div>
          <hr
            class="ms-fontColor-neutralTertiaryAlt launch-pad-hr"
          />
        </div>
        <div>
          <div
            class="ms-Grid"
          >
            <div
              class="ms-Grid-row"
            >
              <div
                class="ms-Grid-col ms-sm3 popup-start-dialog-icon-circle"
              >
                <i
                  aria-hidden="true"
                  class="popup-start-dialog-icon root-44"
                  data-icon-name="testBeaker"
                  role="presentation"
                >
                  
                </i>
              </div>
              <div
                class="ms-Grid-col ms-sm9"
              >
                <div
                  class="launch-pad-item-title"
                >
                  <button
                    aria-describedby="launch-pad-item-description-2"
                    class="ms-Link root-42"
                    role="link"
                  >
                    Assessment
                  </button>
                </div>
                <div
                  class="launch-pad-item-description"
                  id="launch-pad-item-description-2"
                >
                  Walk through a guided process for assessing accessibility compliance.
                </div>
              </div>
            </div>
          </div>
          <hr
            class="ms-fontColor-neutralTertiaryAlt launch-pad-hr"
          />
        </div>
        <div>
          <div
            class="ms-Grid"
          >
            <div
              class="ms-Grid-row"
            >
              <div
                class="ms-Grid-col ms-sm3 popup-start-dialog-icon-circle"
              >
                <i
                  aria-hidden="true"
                  class="popup-start-dialog-icon root-44"
                  data-icon-name="Medical"
                  role="presentation"
                >
                  
                </i>
              </div>
              <div
                class="ms-Grid-col ms-sm9"
              >
                <div
                  class="launch-pad-item-title"
                >
                  <button
                    aria-describedby="launch-pad-item-description-3"
                    class="ms-Link root-42"
                    role="link"
                  >
                    Ad hoc tools
                  </button>
                </div>
                <div
                  class="launch-pad-item-description"
                  id="launch-pad-item-description-3"
                >
                  Get quick access to all automated and assisted checks.
                </div>
              </div>
            </div>
          </div>
          <hr
            class="ms-fontColor-neutralTertiaryAlt launch-pad-hr"
          />
        </div>
      </div>
    </main>
    <div
      class="launch-pad-footer"
      role="complementary"
    >
      <div>
        Version 1.0.4  |  Powered by 
        <a
          class="ms-Link insights-link root-53"
          href="https://www.npmjs.com/package/axe-core"
          target="_blank"
          title="Navigate to axe-core npm page"
        >
          axe-core
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
